File: Analizador_Semantico_Cpp/analyzer.cpp
/*
 * Analizador semántico MUY simplificado en C++
 * ------------------------------------------------
 * - Tipos soportados: int, float
 * - Variables globales, locales y funciones.
 * - Reglas:
 *     · variable usada debe estar declarada
 *     · función llamada debe existir
 *     · número y tipo de argumentos debe coincidir
 *     · advertencia cuando se asigna float -> int
 *
 * Compilación:
 *     g++ -std=c++17 analyzer.cpp -o analyzer
 *
 * Uso:
 *     ./analyzer ejemplo.c
 *
 * Este código NO es un compilador real, solo un chequeo
 * lineal con expresiones muy básicas (suma y llamadas).
 */
#include <iostream>
#include <fstream>
#include <regex>
#include <string>
#include <unordered_map>
#include <vector>
#include <sstream>
#include <algorithm>

struct Function {
    char ret;
    std::vector<char> params;  // tipos de parámetros
};

// Helpers ----------------------------------------------------
bool isFloatLiteral(const std::string &tok) {
    return std::regex_match(tok, std::regex(R"(^\d+\.\d+$)"));
}
bool isIntLiteral(const std::string &tok) {
    return std::regex_match(tok, std::regex(R"(^\d+$)"));
}
char literalType(const std::string &tok) {
    if (isFloatLiteral(tok)) return 'f';
    if (isIntLiteral(tok))   return 'i';
    return 0;
}
std::vector<std::string> split(const std::string &s, char delim) {
    std::vector<std::string> out;
    std::stringstream ss(s);
    std::string item;
    while (std::getline(ss, item, delim)) {
        item.erase(item.begin(), std::find_if(item.begin(), item.end(),
                     [](unsigned char ch){ return !std::isspace(ch); }));
        item.erase(std::find_if(item.rbegin(), item.rend(),
                     [](unsigned char ch){ return !std::isspace(ch); }).base(), item.end());
        if (!item.empty()) out.push_back(item);
    }
    return out;
}
char tokenType(const std::string &tok,
               const std::unordered_map<std::string,char> &localVars,
               const std::unordered_map<std::string,char> &globalVars)
{
    if (auto it = localVars.find(tok); it != localVars.end()) return it->second;
    if (auto it = globalVars.find(tok); it != globalVars.end()) return it->second;
    char lit = literalType(tok);
    if (lit) return lit;
    return 0; // undefined
}

// Expression evaluation (only + or function call or literal/var)
char evalExpr(const std::string &expr,
              const std::unordered_map<std::string,char> &localVars,
              const std::unordered_map<std::string,char> &globalVars,
              const std::unordered_map<std::string,Function> &funcs,
              std::vector<std::string> &errors,
              const std::string &ctx)
{
    std::string e = expr;
    e.erase(remove_if(e.begin(), e.end(), ::isspace), e.end()); // remove spaces

    // function call f(...)
    std::smatch m;
    if (std::regex_match(e, m, std::regex(R"(^(\w+)\((.*)\)$)"))) {
        std::string name = m[1];
        std::string argstr = m[2];
        auto fit = funcs.find(name);
        if (fit == funcs.end()) {
            errors.push_back("[" + ctx + "] función '" + name + "' no declarada");
            return 0;
        }
        const Function &fn = fit->second;
        std::vector<std::string> argTokens = split(argstr, ',');
        if (argTokens.size() != fn.params.size()) {
            errors.push_back("[" + ctx + "] número de argumentos incorrecto en '" + name + "'");
        }
        size_t n = std::min(argTokens.size(), fn.params.size());
        for (size_t i=0;i<n;++i) {
            char at = tokenType(argTokens[i], localVars, globalVars);
            if (at == 0) errors.push_back("[" + ctx + "] identificador '" + argTokens[i] + "' no declarado");
            else if (at != fn.params[i]) {
                errors.push_back("[" + ctx + "] tipo arg '" + argTokens[i] + "' (" +
                                 std::string(1,at)+") esperado "+ std::string(1,fn.params[i]));
            }
        }
        return fn.ret;
    }

    // addition a+b
    size_t plusPos = e.find('+');
    if (plusPos != std::string::npos) {
        std::string left = e.substr(0, plusPos);
        std::string right = e.substr(plusPos+1);
        char lt = tokenType(left, localVars, globalVars);
        char rt = tokenType(right, localVars, globalVars);
        return (lt=='f' || rt=='f') ? 'f' : 'i';
    }

    // variable or literal
    return tokenType(e, localVars, globalVars);
}

// ----------------------------------------------------------------
int main(int argc, char* argv[]) {
    if (argc != 2) {
        std::cout << "Uso: " << argv[0] << " <archivo.c>\n";
        return 1;
    }
    std::ifstream in(argv[1]);
    if (!in) {
        std::cerr << "No se puede abrir archivo\n";
        return 1;
    }

    std::string line;
    std::vector<std::string> lines;
    while (std::getline(in, line)) {
        // eliminar comentarios y espacios
        size_t pos = line.find("//");
        if (pos != std::string::npos) line = line.substr(0,pos);
        std::string trimmed = std::regex_replace(line, std::regex("^\s+|\s+$"), "");
        if (!trimmed.empty())
            lines.push_back(trimmed);
    }

    std::unordered_map<std::string,char> globalVars;
    std::unordered_map<std::string,Function> funcs;
    std::vector<std::string> errors, warns;

    for (size_t i=0;i<lines.size();) {
        std::string l = lines[i];

        // function definition
        std::smatch mfn;
        if (std::regex_match(l, mfn, std::regex(R"((int|float|void)\s+(\w+)\s*\(([^)]*)\)\s*\{)"))) {
            char retType = mfn[1]=="int"? 'i': (mfn[1]=="float"? 'f':'v');
            std::string fname = mfn[2];
            std::string paramsRaw = mfn[3];

            std::unordered_map<std::string,char> localVars;
            Function fn; fn.ret = retType;

            if (!paramsRaw.empty()) {
                std::vector<std::string> paramParts = split(paramsRaw, ',');
                for (std::string p: paramParts) {
                    std::smatch mp;
                    if (std::regex_match(p, mp, std::regex(R"(\s*(int|float)\s+(\w+)\s*)"))) {
                        char pt = mp[1]=="int"? 'i':'f';
                        std::string pname = mp[2];
                        localVars[pname]=pt;
                        fn.params.push_back(pt);
                    }
                }
            }
            funcs[fname]=fn;

            // read body until }
            ++i;
            while (i<lines.size() && lines[i]!="}") {
                std::string b = lines[i];

                // local var decl int x; float y, z;
                std::smatch mvar;
                if (std::regex_match(b, mvar, std::regex(R"(\s*(int|float)\s+([^;]+);)"))) {
                    char vtype = mvar[1]=="int"? 'i':'f';
                    std::vector<std::string> ids = split(mvar[2], ',');
                    for (auto id: ids) {
                        localVars[id]=vtype;
                    }
                } else {
                    // assignment
                    std::smatch masg;
                    if (std::regex_match(b, masg, std::regex(R"(\s*(\w+)\s*=\s*(.+);)"))) {
                        std::string lhs = masg[1];
                        std::string rhs = masg[2];
                        char lhs_t = tokenType(lhs, localVars, globalVars);
                        if (lhs_t==0) {
                            errors.push_back("["+fname+"] variable '"+lhs+"' no declarada");
                            lhs_t='u';
                        }
                        char rhs_t = evalExpr(rhs, localVars, globalVars, funcs, errors, fname);
                        if (lhs_t!=rhs_t && rhs_t!=0) {
                            warns.push_back("["+fname+"] asignación "+
                                            std::string(1,rhs_t)+"->"+std::string(1,lhs_t)+
                                            " en '"+lhs+"'");
                        }
                    }
                }
                ++i;
            }
            ++i; // skip }
            continue;
        }

        // global variable declaration
        std::smatch mvar;
        if (std::regex_match(l, mvar, std::regex(R"(\s*(int|float)\s+([^;]+);)"))) {
            char gtype = mvar[1]=="int"? 'i':'f';
            std::vector<std::string> ids = split(mvar[2], ',');
            for (auto id: ids) globalVars[id]=gtype;
        }
        ++i;
    }

    // ----------- output -----------------
    if (errors.empty()) {
        std::cout << "\u2705 Sin errores semánticos\n";
    } else {
        for (auto &e: errors) std::cout << "\u274C " << e << "\n";
    }
    for (auto &w: warns) std::cout << "\u26A0\uFE0F " << w << "\n";
    return 0;
}

File: ANALYZER/analyzer.py
"""
Analizador semántico MUY sencillo para un subconjunto de C:
· Tipos soportados: int, float
· Declara variables y funciones
· Detecta: variables no declaradas, funciones no declaradas,
  número/tipo de argumentos y asignaciones int↔float

Uso:
    python analyzer.py ejemplo.c
"""

import re, sys, textwrap

# ---------- utilidades de análisis --------------------------------
def parse_decl(line):
    """
    Reconoce 'int a;'  o  'float x, y;'
    Devuelve (tipo, [nombres])  o  (None, [])
    """
    m = re.match(r'\s*(int|float)\s+([^;]+);', line)
    if not m:
        return None, []
    typ = m.group(1)
    names = [n.strip() for n in m.group(2).split(',')]
    return typ, names

def literal_type(tok):
    if re.fullmatch(r'[0-9]+\.[0-9]+', tok):
        return 'float'
    if re.fullmatch(r'[0-9]+', tok):
        return 'int'
    return None

# ---------- analizador principal ----------------------------------
def analyze(code:str):
    lines = [l.strip() for l in code.splitlines() if l.strip() and not l.strip().startswith('//')]

    global_vars   = {}        # nombre → tipo
    functions     = {}        # nombre → {'return': tipo, 'params': [(nombre,tipo),...]}
    errors, warns = [], []

    i = 0
    while i < len(lines):
        line = lines[i]

        # --------- declaración de función -----------
        m_fun = re.match(r'(int|float|void)\s+(\w+)\s*\(([^)]*)\)\s*\{', line)
        if m_fun:
            ret_type, fn_name, params_raw = m_fun.groups()
            local = {}  # tabla de símbolos local
            params = []

            if params_raw.strip():
                for p in params_raw.split(','):
                    t, names = parse_decl(p + ';')  # reuse
                    for n in names:
                        params.append((n, t))
                        local[n] = t
            functions[fn_name] = {'return': ret_type, 'params': params}

            # leer cuerpo
            i += 1
            while i < len(lines) and lines[i] != '}':
                inner = lines[i]

                # variables locales
                t, names = parse_decl(inner)
                if t:
                    for n in names:
                        local[n] = t
                    i += 1
                    continue

                # asignaciones
                m_asg = re.match(r'(\w+)\s*=\s*([^;]+);', inner)
                if m_asg:
                    lhs, rhs = m_asg.groups()
                    lhs_t = local.get(lhs) or global_vars.get(lhs)
                    if lhs_t is None:
                        errors.append(f"[{fn_name}] variable '{lhs}' no declarada")
                        lhs_t = 'undef'

                    rhs_t = eval_expr(rhs, local, global_vars, functions, errors, fn_name)

                    if lhs_t != rhs_t and rhs_t not in ('undef',):
                        warns.append(f"[{fn_name}] asignación {rhs_t}->{lhs_t} en '{lhs}'")
                i += 1
            # saltar la llave de cierre
            i += 1
            continue

        # --------- declaración global de variable -----------
        t, names = parse_decl(line)
        if t:
            for n in names:
                global_vars[n] = t
        i += 1

    return errors, warns

# ---------- evaluación de expresiones muy simple ------------------
def eval_expr(expr, local, global_, fns, errors, ctx):
    expr = expr.strip()

    # llamada de función  f(x,y)
    m_call = re.match(r'(\w+)\((.*)\)', expr)
    if m_call:
        name, argstr = m_call.groups()
        if name not in fns:
            errors.append(f"[{ctx}] función '{name}' no declarada")
            return 'undef'
        fn = fns[name]
        arg_tokens = [a.strip() for a in argstr.split(',')] if argstr.strip() else []
        if len(arg_tokens) != len(fn['params']):
            errors.append(f"[{ctx}] número de argumentos incorrecto en '{name}'")
        # comprobar tipos
        for tok, (_pname, ptype) in zip(arg_tokens, fn['params']):
            atype = token_type(tok, local, global_)
            if atype != ptype:
                errors.append(f"[{ctx}] tipo arg '{tok}' ({atype}) esperado {ptype}")
        return fn['return']

    # suma  a+b
    m_sum = re.match(r'(.+)\+(.+)', expr)
    if m_sum:
        lt = token_type(m_sum.group(1).strip(), local, global_)
        rt = token_type(m_sum.group(2).strip(), local, global_)
        return 'float' if 'float' in (lt, rt) else 'int'

    # literal o variable
    return token_type(expr, local, global_)

def token_type(tok, local, global_):
    return local.get(tok) or global_.get(tok) or literal_type(tok) or 'undef'

# ---------- CLI ---------------------------------------------------
def main():
    if len(sys.argv) != 2:
        print("Uso: python analyzer.py <archivo.c>")
        return
    with open(sys.argv[1], encoding='utf-8') as f:
        code = f.read()

    errs, warns = analyze(code)
    if errs:
        for e in errs: print("❌", e)
    else:
        print("✅ Sin errores semánticos")
    for w in warns: print("⚠️", w)

if __name__ == "__main__":
    main()
